### PR TITLE
Introduction to Retrofit

### PR DESCRIPTION
**Pull request by ranjanih**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [ ] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [ ] the article explains *why* and not only *how*
- [ ] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [ ] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [ ] the text is conversational ("we" instead of "you", easy language)
- [ ] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [ ] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [ ] the text uses active voice instead of passive voice
- [ ] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [ ] file names and variables are highlighted as `code`

## Structure
- [ ] the article has a short intro before the first headline
- [ ] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [ ] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [ ] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [ ] code examples are introduced with a ":" and explained below the code example
- [ ] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: content/blog/2022/2022-04-26-introduction-to-retrofit.md [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+````
+With `Gradle`:
+````groovy
+        dependencies {  
+            implementation 'com.squareup.retrofit2:retrofit:2.5.0'
+            implementation 'com.squareup.retrofit2:converter-jackson:2.5.0'
+        }
+````
+
+## Setting Up a Retrofit Client
```
Comment (thombergs) [inline content/blog/2022/2022-04-26-introduction-to-retrofit.md:90]: It took a while until I understood that this section is about "quickly" showing what steps are needed. We should rename this section to "Quick Guide to Setting up a Retrofit Client" and mention in the introduction that we're going to look at the details in the upcoming section. I was confused as to why things were repeated in the next section.

File: content/blog/2022/2022-04-26-introduction-to-retrofit.md [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+````
+
+## Using Retrofit
+In the further sections we will learn more about the Retrofit API and how to use them. 
+
+### Building a Client Interface
+In this section, we will look at how to build the client interface.
+**Retrofit supports annotations @GET, @POST, @PUT, @DELETE, @PATCH, @OPTIONS, @HEAD** which we use to annotate our client methods as shown below
+
+````java
```
Comment (thombergs) [inline content/blog/2022/2022-04-26-introduction-to-retrofit.md:207]: Please make each group of the examples its own section (except the first, maybe), even if that section hast only one paragraph. I.e.

* Path parameters
* Query parameters
* Multiple Query Parameters
* Request body
* Headers
*

File: content/blog/2022/2022-04-26-introduction-to-retrofit.md [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder()
+                .connectTimeout(props.getConnectionTimeout(), TimeUnit.SECONDS)
+                .readTimeout(props.getReadWriteTimeout(), TimeUnit.SECONDS);
+
+        return new Retrofit.Builder().client(httpClientBuilder.build())
+                .baseUrl(props.getEndpoint())
+                .addConverterFactory(JacksonConverterFactory.create(new ObjectMapper()))
+                .build().create(LibraryClient.class);
+````
+Here, the timeout values are as specified in application.yaml.
```
Comment (thombergs) [inline content/blog/2022/2022-04-26-introduction-to-retrofit.md:280]: Like I mentioned above, please provide a code example of the relevant part of the application.yml (in the code example of my previous comment about this, maybe just point to this place here)

File: content/blog/2022/2022-04-26-introduction-to-retrofit.md [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+````java
+    @Test
+    @DisplayName("Successful getAllBooks call")
+    public void getAllBooksTest() throws Exception {
+        LibraryAuditService libraryAuditService = applicationContext.getBean(LibraryAuditService.class);
+        String booksResponse = getBooksResponse("/response/getAllBooks.json");
+        List<BookDto> bookDtoList =
+                new ObjectMapper().readValue(booksResponse, new TypeReference<>() {
+                });
+        when(libraryClient.getAllBooks("all")).thenReturn(Calls.response(bookDtoList));
```
Comment (thombergs) [inline content/blog/2022/2022-04-26-introduction-to-retrofit.md:519]: The piece of code is missing where the `libraryClient` is initialized as a Mockito mock (i.e. something like `final LibraryClient libraryClient = Mockito.mock(LibraryClient.class)`. Without this info, you leave the reader guessing :).

Also, there is too much going on in this code example when all you want to show is how to mock the library client (including Spring Boot in this test is too much, for example).

**Review by thombergs — CHANGES_REQUESTED** [review]
Hey @ranjanih Please have a look through my comments. We'll probably need a 2nd pass (with fewer comments) after this one, but we're going to make this a pretty cool article!

File: content/blog/2022/2022-04-26-introduction-to-retrofit.md [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+- Support for synchronous and asynchronous calls
+
+## What is `Retrofit`
+[Retrofit](https://square.github.io/retrofit/) is a high-level REST abstraction built on top of OkHttp. 
+When used to call REST applications, it greatly simplifies API interactions by parsing requests and responses into POJOs.
+
+In the further sections, we will work on creating a Retrofit client and look at how to incorporate the various features that OkHttp provides.
+
+## Setting up an existing REST Service Application
+We will use a sample REST-based [Spring Boot Library Application](https://github.com/ranjanih/code-examples/tree/ranjani-retrofit/core-java/retrofit/introduction-to-retrofit/SimpleLibraryApplication) that acts as a REST service.
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-04-26-introduction-to-retrofit.md]: I meant it in the context of library where we talk about books and authors. I have updated the wording to make it more clear.

File: content/blog/2022/2022-04-26-introduction-to-retrofit.md (hunk 1 above)
Comment (ranjanih) [reply content/blog/2022/2022-04-26-introduction-to-retrofit.md]: Makes sense. Updated.

File: content/blog/2022/2022-04-26-introduction-to-retrofit.md [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+This library application is a Spring Boot service that uses Maven for build and HSQLDB as the underlying database.
+The Maven Wrapper bundled with the application will be used to start the service:
+````text
+    mvnw clean verify spring-boot:run (for Windows)
+    ./mvnw clean verify spring-boot:run (for Linux)
+````
+Now, the application should successfully start:
+{{% image alt="settings" src="images/posts/retrofit/success_at_startup.jpg" %}}
+
+The Swagger documentation of this application can be viewed at
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-04-26-introduction-to-retrofit.md]: Add a brief overview about Swagger before i use it in the subsequent sections.

**Pull request merged**